apps/sync/js/sync-engine/syncengine.js

```diff
@@ -0,0 +1,274 @@
+        }
+      }).catch(err => {
+        if (err instanceof TypeError) {
+          throw new SyncEngine.UnrecoverableError();
+        } else if (err instanceof Error && typeof err.request === 'object') {
+          if (err.request.status === 401) {
+            throw new SyncEngine.AuthError();
+          } else {
+            throw new SyncEngine.TryLaterError();
+          }
+        } else {
+          throw new SyncEngine.UnrecoverableError();
+        }
+      });
+    },
+
+    _storageVersionOK: function(metaGlobal) {
+      var payloadObj;
+      try {
+        payloadObj = JSON.parse(metaGlobal.data.payload);
+      } catch(e) {
+        return false;
+      }
+      return (typeof payloadObj === 'object' &&
+          payloadObj.storageVersion === 5);
+    },
+
+    _initFxSyncWebCrypto: function(cryptoKeys) {
+      this._fswc = new FxSyncWebCrypto();
+      return this._fswc.setKeys(this._kB, cryptoKeys).then(() => {}, (err) => {
+        if (err === 'SyncKeys hmac could not be verified with current main ' +
+            'key') {
+          throw new SyncEngine.UnrecoverableError();
+        } else {
+          throw err;
+        }
+      });
+    },
+
+    connect: function() {
+      return this._syncCollection('meta', true)
+      .then(() => this._getItem('meta', 'global', true))
+      .then(metaGlobal => {
+        if (!this._storageVersionOK(metaGlobal)) {
+          return Promise.reject(new SyncEngine.UnrecoverableError(`Incompatible\
+ storage version or storage version not recognized.`));
+        }
+        return this._syncCollection('crypto', true);
+      }).then(() => this._getItem('crypto', 'keys', true))
+      .then((cryptoKeysRecord) => {
+        var cryptoKeys;
+        try {
+          cryptoKeys = JSON.parse(cryptoKeysRecord.data.payload);
+        } catch (e) {
+          return Promise.reject(new SyncEngine.UnrecoverableError(`Could not pa\
+rse crypto/keys payload as JSON`));
+        }
+        return cryptoKeys;
+      }).then((cryptoKeys) => {
+        //Cannot do this from the constructor because we need cryptoKeys first,
+        //and it's not nice to return a promise from a constructor:
+        return this._initFxSyncWebCrypto(cryptoKeys);
+      });
+    },
+
+    registerAdapter: function(collectionName, adapter) {
+      this._collections[collectionName] = this._kinto.collection(
+          collectionName);
+      this._collections[collectionName].use(new FxSyncIdSchema(
+          collectionName));
+      this._collections[collectionName].use(new WebCryptoTransformer(
+          collectionName, this._fswc));
+      this._adapters[collectionName] = adapter;
+    },
+
+    _updateCollection: function(collectionName) {
+      return this._syncCollection(collectionName)
+      .then(() => this._adapters[collectionName].update(
+          this._collections[collectionName]))
+      .then(() => this._syncCollection(collectionName));
+    },
+
+    syncNow: function() {
```

Comment:
Note that we won't always be synchronizing all collections. The user can choose not to sync some specific collections via Settings app.